File: users/clusteringUsersFixed.py
# Note that the add operation in K-means algorithm is defined WITHOUT factors

import json
import os
import sys
from operator import add
from pyspark import SparkContext

import numpy as np
from scipy.sparse import csc_matrix
# from scipy.sparse import dok_matrix

def gen_rdd(tweets_file):
    return sc.textFile(tweets_file).map(process_tweet)


def process_tweet(raw_data):
    try:
        raw_data = raw_data.split('\t')
        raw_json = json.loads(raw_data[3])
        uid = raw_json['user']['id']
        text = [w.strip() for w in raw_data[0].split()]
        return (uid, (1, text))
    except:
        return (0, (0, []))


def clustering_user(files):

    w_dict = None
    w_len = 0
    kPoints = None

    sqsum = lambda v: np.sum(v.data * v.data)

    def normalize(user):
        uid = user[0]
        counter = float(user[1][0])
        tokens = sorted(user[1][1])

        vec = []
        tp = (tokens[0], 1)
        for i in range(1,len(tokens)):
            if tp[0] == tokens[i]:
                tp = (tp[0], tp[1] + 1)
            else:
                vec.append(tp)
                tp = (tokens[i], 1)
        vec.append(tp)
        vec = map(lambda (token, weight): (token, weight / counter), vec)
        return (uid, vec)


    def to_sparse(user):
        uid = user[0]
        vec = user[1]

        row = []
        data = []
        for i in range(len(vec)):
            idx = None
            try:
                idx = w_dict[vec[i][0]]
            except:
                # Do nothing: filter out words appears infrequently
                pass
            if idx:
                row.append(idx)
                data.append(vec[i][1])

        return (uid, csc_matrix((data, (row, [0] * len(row))), shape=(w_len, 1)))


    def closestPoint(user):
        vec = user[1]
        bestIndex = 0
        closest = float("+inf")
        for i in range(len(kPoints)):
            tempDist = sqsum(vec - kPoints[i])
            if tempDist < closest:
                closest = tempDist
                bestIndex = i
        # return (bestIndex, (vec, 1, user[0]))
        return (bestIndex, (vec, 1))


    if len(files) == 0:
        return 0

    # Ignore users that have fewer than MIN_POST tweets
    MIN_POST = 7
    MIN_OCCURS = 10000
    # MIN_POST = 0
    # MIN_OCCURS = 0

    # For each account, put all the tokens of its tweets into one vector
    all_users = None
    for file_name in files:
        users = gen_rdd(file_name)
        if all_users:
            all_users = all_users.union(users)
        else:
            all_users = users

    # Normalize the vector
    all_users = all_users.reduceByKey(lambda a, b: (a[0] + b[0], a[1] + b[1]), numPartitions = 480) \
            .filter(lambda (uid, (counter, tokens)): counter >= MIN_POST)

    # 4606200 words
    all_words = all_users.flatMap(lambda (uid, (counter, tokens)): [(w, 1) for w in tokens]) \
            .reduceByKey(add, numPartitions = 240) \
            .filter(lambda (w, count): count >= MIN_OCCURS) \
            .map(lambda (w, count): w).collect()
    all_words = sorted(all_words)
    w_len = len(all_words)
    w_dict = {all_words[i]: i for i in range(len(all_words))}

    print w_len

    # Construct features
    all_users = all_users.map(normalize).map(to_sparse).cache()

    # Clustering (k-means)
    K = 30
    CONVERGE = 0.1
    MAXITER = 10
    temp_dist = 1.0
    kPoints = [vec for (uid, vec) in all_users.takeSample(False, K, 1)]

    iter = 0
    while temp_dist > CONVERGE and iter < MAXITER:
        closest = all_users.map(closestPoint)
        pointStats = closest.reduceByKey(
            lambda (vec1, counter1), (vec2, counter2): (vec1 + vec2, counter1 + counter2), \
            numPartitions = 480)
        newPoints = pointStats.map(
            lambda (cid, (vec, counter)): (cid, vec / counter)).collect()

        temp_dist = np.sum([sqsum(kPoints[cid] - vec) for (cid, vec) in newPoints])
        print "Iteration %d:" % (iter + 1), temp_dist

        for (cid, vec) in newPoints:
            kPoints[cid] = vec

        iter = iter + 1

    print "Final centers:"
    for point in kPoints:
        print "=========="
        row, col = point.nonzero()
        for i,j in zip(row, col):
            print i,j,point[i,j]
        print "=========="
    print "w_len =", w_len

    # print '\n\nSize of each group:'
    # closest = all_users.map(closestPoint).cache()
    # for i in range(30):
    #     print "Group %02d" % i
    #     print [t[1][2] for t in closest.filter(lambda (index, others): index == i).collect()]

    return all_users.count()
 

if __name__ == "__main__":
    reload(sys)
    sys.setdefaultencoding('utf8')
    sc = SparkContext("spark://ion-21-14.sdsc.edu:7077", "ClusteringUsersPartitions", pyFiles=['clusteringUsersFixed.py'])

    dir_path = '/user/arapat/twitter-tag/'
    # files = ['/user/arapat/twitter-sample/tag100']
    # files = [dir_path + 't01']
    # files = [dir_path + 't%02d' % k for k in range(1, 11)]
    files = [dir_path + 't%02d' % k for k in range(1, 71)] + [dir_path + 'u%02d' % k for k in range(1,86)]
    print " ".join(["All users:", str(clustering_user(files))])

File: users/scala/clusteringUsersFixed.scala

import breeze.linalg.{Vector, SparseVector, squaredDistance}

import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.json4s._
import org.json4s.native.JsonMethods


object SparkKMeans {

    def genRdd(tweetsFile: String, sc: SparkContext) = {
        sc.textFile(tweetsFile).map(processTweet)
    }

    def processTweet(rawData: String) : (Int, Array[String]) = {
        try {
            rawData = raw_data.split('\t')
            val rawJson = JsonMethods.parse(rawData(3))
            val uid = (rawJson \ "user" \ "id").values
            val tokens = rawData(0).split(' ').map(_.trim)
            return (uid, tokens)
        } catch {
            case _ => // broken tweets
                return (0, Array())
        }
    }

    def normalize(user: (Int, (Int, Array[String]))) = {
        val tokenSet = user.tokens.toSet
        val tokenPairs = tokenSet.map(x => (x, user.tokens.count(_ == x)))
            .map(tp => (tp._1, tp._2 / user.counter.toFloat)).toArray
        (user.uid, tokenPairs)
    }

    def toSparse(user: (Int, Array[String]), dict: Map[String, Int]) = {
        val seq = user.tokens
            .map(tp => (dict.get(tp._1), tp._2))
            .filter(elm => elm._1 != None).sorted
        val vector = SparseVector(
            seq.map(elm => elm._1), seq.map(elm => elm._2), dict.size)
        (user.uid, vector)
    }

    def closestPoint(p: SparseVector[Double], centers: Array[Vector[Double]]): Int = {
        var index = 0
        var bestIndex = 0
        var closest = Double.PositiveInfinity

        for (i <- 0 until centers.length) {
            val tempDist = squaredDistance(p, centers(i))
            if (tempDist < closest) {
                closest = tempDist
                bestIndex = i
            }
        }

        bestIndex
    }

    def clusteringUser(sc: SparkContext, files: Array[String]) {
        if (files.length == 0)
            return

        // Ignore users that have fewer than MIN_POST tweets
        val MIN_POST = 7
        val MIN_OCCURS = 10000
        // val MIN_POST = 0
        // val MIN_OCCURS = 0

        val rawUsers = files.map(genRdd(_, sc))
        var allUsers = rawUsers(0)
        for (i <- 1 until rawUsers.length) {
            allUsers = allUsers.union(rawUsers(i)) 
        }

        // Normalize the vector
        allUsers = allUsers.map((uid, tokens) => (uid, (1, tokens)))
            .reduceByKey {case ((x1, y1), (x2, y2)) => (x1 + x2, y1 + y2)}
            .filter {case (uid, (counter, tokens)) => counter >= MIN_POST}

        allWords = allUsers.flatMap {case (uid, (counter, tokens)) => tokens.map((_, 1))}
            .reduceByKey(_ + _)
            .filter {case (token, counter) => counter >= MIN_OCCURS}
            .map {case (token, counter) => token}
            .collect()
        allWords = allWords.sorted
        dict = {for (i <- 0 until allWords.length) yield (i, allWords(i))}.toMap

        // Construct features
        allUsers = allUsers.map(normalize).map(toSparse).cache()

        // Clustering (k-means)
        val K = 30
        val CONVERGE = 0.1
        val MAX_ITER = 10
        var tempDist = 1.0
        val kPoints = allUsers.takeSample(withReplacement = false, K, 42)
            .map {case (uid, vector) => vector}
            .toArray

        var iter = 0
        while(tempDist > convergeDist && iter < MAX_ITER) {

            val closest = allUsers.map {case (uid, p) => (closestPoint(p, kPoints), (p, 1))}

            val pointStats = closest.reduceByKey {case ((x1, y1), (x2, y2)) => (x1 + x2, y1 + y2)}

            val newPoints = pointStats.map {pair =>
                (pair._1, pair._2._1 * (1.0 / pair._2._2))}.collectAsMap()

            tempDist = 0.0 
            for (i <- 0 until K) {
                tempDist += squaredDistance(kPoints(i) - newPoints(i))
            }

            for (newP <- newPoints) {
                kPoints(newP._1) = newP._2
            }

            iter = iter + 1
            println("Finished iteration " + iter + " (delta = " + tempDist + ")")
        }

        println("Final centers:")
        kPoints.foreach(println)
        // for point in kPoints:
        // print "=========="
        // row, col = point.nonzero()
        // for i,j in zip(row, col):
        // print i,j,point[i,j]
        // print "=========="
        // print "w_len =", w_len

        // print '\n\nSize of each group:'
        // closest = all_users.map(closestPoint).cache()
        // for i in range(30):
        //     print "Group %02d" % i
        //     print [t[1][2] for t in closest.filter(lambda (index, others): index == i).collect()]

        // return all_users.count()
    }


    def main(args: Array[String]) {
        val sc = new SparkContext("spark://ion-21-14.sdsc.edu:7077", "ClusteringUsers",
              System.getenv("SPARK_HOME"), SparkContext.jarOfClass(this.getClass).toSeq)
        
        val dirPath = "/user/arapat/twitter-tag/"
        // val files =
        //     {for (i <- 1 to 70) yield dirPath + "t%02d".format(i)}
        //     ++ {for (i <- 1 to 85) yield dirPath + "u%02d".format(i)}
        val files = Array(dirPath + "t02d".format(1))
        clusteringUser(sc, files)
    }
}

